File: PYTHON/5-BinarySearch/2-search_matrix.py
from typing import List
import bisect

def searchMatrix(matrix: List[List[int]], target: int) -> bool:
    firstColumn = [line[0] for line in matrix]
    idxLine = bisect.bisect_right(firstColumn, target) - 1

    if idxLine < 0:
        return False
    
    lineTarg = matrix[idxLine]
    idxCol = bisect.bisect_left(lineTarg, target)

    return idxCol < len(lineTarg) and lineTarg[idxCol] == target

# # brute force -> time complexity: O(m * n), space complexity: O(1)
# def searchMatrix(matrix: List[List[int]], target: int) -> bool:
#     for i in range(len(matrix)):
#         for j in range(len(matrix[0])):
#             if matrix[i][j] == target:
#                 return True
#     return False

# # binary search one pass
# def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
#     ROWS, COLS = len(matrix), len(matrix[0])

#     l, r = 0, ROWS * COLS - 1
#     while l <= r:
#         m = l + (r - l) // 2
#         row, col = m // COLS, m % COLS
#
#         if target > matrix[row][col]:
#             l = m + 1
#         elif target < matrix[row][col]:
#             r = m - 1
#         else:
#             return True
#     return False

if __name__ == "__main__":
    matrix = [[1,2,4,8],[10,11,12,13],[14,20,30,40]]
    target = 10
    print(searchMatrix(matrix, target))

File: CPP/5-BinarySearch/2-search_matrix.cpp
#include <iostream>
#include <vector>

using namespace std;

// brute force
// bool searchMatrix(vector<vector<int>>& matrix, int target) {
//     int row = matrix.size();
//     int col = matrix[0].size();

//     for (int i = 0; i < row; i++) {
//         for (int j = 0; j < col; j++) {
//             if (target == matrix[i][j]) {
//                 return true;
//             }
//         }
//     }

//     return false;
// }

// melhor
bool searchMatrix(vector<vector<int>>& matrix, int target) {
    int rows = matrix.size();
    int cols = matrix[0].size();
    
    int left = 0;
    int right = rows * cols - 1;

    while (left <= right) {
        int mid = (right + left) / 2;
        int row = mid / cols;
        int col = mid % cols;
        if (target > matrix[row][col]) {
            left = mid + 1;
        } else if (target < matrix[row][col]) {
            right = mid - 1;
        } else {
            return true;
        }
    }

    return false;
}

int main() {
    vector<vector<int>> matrix = {
        {1, 2, 4, 8}, 
        {10, 11, 12, 13},  
        {14, 20, 30, 40}
    };
    int target = 10;
    
    cout << searchMatrix(matrix, target);
}